### PR TITLE
feature: introduce [$ dune ocaml doc]

### PR DESCRIPTION
**Pull request by EmileTrotignon**
This is a feature from issue #6831

It adds a new command "dune ocaml doc" that builds the doc, and then open its index in a browser. If no browser can be found, it print the location of the index.

There are no tests for this feature, as I do not know of a way to automatically test it.

I also only tested this on my machine, and this does very different stuff on different systems, so I would like if someone with a windows or an osx machine would be able to test it.

### CODE REVIEW

**Comment by Alizter** [comment]
Can you setup a cram test showing the feature. It should be possible to mock a browser for xdg-open in the cram test.

**Comment by EmileTrotignon** [comment]
@Alizter I have not managed to mock a browser for xdg-open, so I justed mock xdg-open itself.

File: boot/libs.ml
Comment (Alizter) [inline boot/libs.ml:1]: You probably want to revert this change. This happens automatically I think on macos.

File: test/blackbox-tests/test-cases/doc-browser.t/run.t [hunk 1]
```diff
@@ -0,0 +1,12 @@
+This tests shows how to use the `dune ocaml doc` command to open the
+documentation index to a browser.
+  $ ls
+  dune
+  dune-project
+  foo.ml
+  spoof.desktop
+  xdg-open
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/doc-browser.t/run.t:8]: You can remove this as this is just noise.

File: test/blackbox-tests/test-cases/doc-browser.t/run.t [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  $ PATH=.:$PATH 
+  $ export PATH
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/doc-browser.t/run.t:10]:
```suggestion
  $ export PATH=.:$PATH 
```

File: test/blackbox-tests/test-cases/doc-browser.t/run.t [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  $ dune ocaml doc | grep -c "file:///.*/_build/.sandbox/.*/default/test/blackbox-tests/test-cases/doc-browser.t/_build/default/_doc/_html/index.html"
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/doc-browser.t/run.t:11]:
```suggestion
  $ dune ocaml doc | sed -e 's/.*file:\/\/\([^ ]*\).*/\1/'
  $TESTCASE_ROOT/_build/default/_doc/_html/index.html
```
Here is a better way to scrub the output. cram tests usually detect and scrub the current directory but because we start with `file:///` it gets confused. Therefore we can manually remove it with sed.

File: src/dune_rules/odoc.mli [hunk 4]
```diff
@@ -2,6 +2,10 @@
 
 open Import
 
+module Paths : sig
+  val toplevel_index : Context.t -> Path.Build.t
```
Comment (Alizter) [inline src/dune_rules/odoc.mli:6]: Is this supposed to be the location of the index file? If so could you add a comment detailing it as such?

File: test/blackbox-tests/test-cases/doc-browser.t/dune-project [hunk 5]
```diff
@@ -0,0 +1,4 @@
+(lang dune 3.6)
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/doc-browser.t/dune-project:1]: We can use lang 3.8 for this.

File: bin/build_cmd.ml
Comment (Alizter) [inline bin/build_cmd.ml:1]: It's a shame that GitHub messes up the syntax highlighting for OCaml here due to the `{| |}` comments. I have reported the issue upstream, but it is unlikely to be fixed anytime soon.

File: bin/build_cmd.ml [hunk 6]
```diff
@@ -215,3 +215,63 @@ let fmt =
     run_build_command ~common ~config ~request
   in
   Cmd.v (Cmd.info "fmt" ~doc ~man ~envs:Common.envs) term
+
+let doc =
+  let doc = "Build the documentation of an OCaml project" in
```
Comment (Alizter) [inline bin/build_cmd.ml:220]:
```suggestion
  let doc = "Build and view the documentation of an OCaml project" in
```

File: bin/build_cmd.ml [hunk 7]
```diff
@@ -215,3 +215,63 @@ let fmt =
...
+  let man =
+    [ `S "DESCRIPTION"
+    ; `P
+        {|$(b,dune ocaml doc) build and open the documention of an ocaml project.|}
```
Comment (Alizter) [inline bin/build_cmd.ml:224]:
```suggestion
        {|$(b,dune ocaml doc) builds and then opens the documention of an OCaml project in the users default browser.|}
```
Or something similar

File: bin/build_cmd.ml [hunk 8]
```diff
@@ -215,3 +215,63 @@ let fmt =
...
+        , (* First element of argv is the name of the command. *)
+          (cmd_name :: args) @ [ url ] )
+      in
+      match cmd with
+      | Some (cmd, args) ->
+        Proc.restore_cwd_and_execve
+          (Path.to_absolute_filename cmd)
+          args ~env:Env.initial
+      | None ->
+        Printf.printf "Docs built. Index can be found here: %s\n"
```
Comment (Alizter) [inline bin/build_cmd.ml:271]: Rather than `Printf.printf` use `Dune_console.print`. It might also be worth printing this message in both cases, whether or not the command was found. In fact, it might be clearer to raise a user error / warning here and say that a web browser command or whatever could not be found / used so that it doesn't mysteriously fail.

**Review by Alizter — COMMENTED** [review]
I know this is still a draft, but some initial comments and suggestions.

File: bin/build_cmd.ml [hunk 9]
```diff
@@ -215,3 +215,70 @@ let fmt =
...
+        [ Pp.textf "Docs built. Index can be found here: %s\n"
+            absolute_toplevel_index_path
+        ];
+
+      let url = Printf.sprintf {|file://%s|} absolute_toplevel_index_path in
+      let cmd =
+        let open Option.O in
+        let path = Env_path.path Env.initial in
+        let* cmd_name, args =
+          match Ocaml_config.system doc_ctx.ocaml_config with
```
Comment (rgrinberg) [inline bin/build_cmd.ml:259]: We now have a `Platform` module that you can use

File: bin/build_cmd.ml [hunk 10]
```diff
@@ -215,3 +215,70 @@ let fmt =
...
+      let toplevel_index_path = Dune_rules.Odoc.Paths.toplevel_index doc_ctx in
+      let absolute_toplevel_index_path =
+        Path.(toplevel_index_path |> build |> to_absolute_filename)
+      in
+      Dune_console.print
+        [ Pp.textf "Docs built. Index can be found here: %s\n"
+            absolute_toplevel_index_path
+        ];
+
+      let url = Printf.sprintf {|file://%s|} absolute_toplevel_index_path in
```
Comment (rgrinberg) [inline bin/build_cmd.ml:254]: Any reason why you don't use normal strings?

Comment (EmileTrotignon) [reply bin/build_cmd.ml:254]: I guess I was worried of having to escape the `//`, but that's not needed, I'll change that.

File: bin/build_cmd.ml (hunk 9 above)
Comment (EmileTrotignon) [reply bin/build_cmd.ml:259]: Oh that's very nice, thanks !

**Comment by EmileTrotignon** [comment]
I have finally tested this on a mac, and it works ! I still have to test it on windows.
I think mac and linux support is enough to get this out of draft though.

File: bin/build_cmd.mli [hunk 11]
```diff
@@ -12,3 +12,5 @@ val runtest_term : unit Term.t
 val build : unit Cmd.t
 
 val fmt : unit Cmd.t
+
+val doc : unit Cmd.t
```
Comment (Alizter) [inline bin/build_cmd.mli:16]: I don't think this is the correct place for this. Rather than putting it here, could you put it in a file called doc.ml and place that file in bin/ocaml next to all the other ocaml subcommands?

File: bin/build_cmd.ml [hunk 12]
```diff
@@ -213,3 +213,70 @@ let fmt =
...
+      in
+      match cmd with
+      | Some (cmd, args) ->
+        Proc.restore_cwd_and_execve
+          (Path.to_absolute_filename cmd)
+          args ~env:Env.initial
+      | None ->
+        Dune_console.print
+          [ Pp.text
+              "No browser could be found, you will have to open the \
```
Comment (Alizter) [inline bin/build_cmd.ml:276]: Please use `Console.print` instead.

File: bin/build_cmd.ml [hunk 13]
```diff
@@ -213,3 +213,70 @@ let fmt =
...
+      let toplevel_index_path = Dune_rules.Odoc.Paths.toplevel_index doc_ctx in
+      let absolute_toplevel_index_path =
+        Path.(toplevel_index_path |> build |> to_absolute_filename)
+      in
+      Dune_console.print
+        [ Pp.textf "Docs built. Index can be found here: %s\n"
+            absolute_toplevel_index_path
+        ];
+
+      let url = Printf.sprintf "file://%s" absolute_toplevel_index_path in
```
Comment (Alizter) [inline bin/build_cmd.ml:252]: String concatenation is perhaps simpler.

**Comment by EmileTrotignon** [comment]
@Alizter The last commit takes your comment into account.

File: test/blackbox-tests/test-cases/doc-browser.t/run.t [hunk 14]
```diff
@@ -0,0 +1,8 @@
+This tests shows how to use the `dune ocaml doc` command to open the
+documentation index to a browser.
+  $ export PATH=.:$PATH 
+  $ dune ocaml doc | sed -e 's/.*file:\/\/\([^ ]*\).*/\1/'
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/doc-browser.t/run.t:4]: You can use a different deliminator rather than `/` for sed so you don't have to escape so many characters. Something like # should work.

File: test/blackbox-tests/test-cases/doc-browser.t/open [hunk 15]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+echo $@
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/doc-browser.t/open:2]: Adding something like
```
echo "Web browser received arg:"
```
might make the test easier to read. Same for the other one.

Better yet have a single copy and create symlinks so we don't have to duplicate.

File: test/blackbox-tests/test-cases/doc-browser.t/dune-project [hunk 16]
```diff
@@ -0,0 +1,4 @@
+(lang dune 3.7)
```
Comment (Alizter) [inline test/blackbox-tests/test-cases/doc-browser.t/dune-project:1]: Let's use 3.8 here.

File: bin/ocaml/doc.ml [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    match cmd with
+    | Some (cmd, args) ->
+      Proc.restore_cwd_and_execve
+        (Path.to_absolute_filename cmd)
+        args ~env:Env.initial
+    | None ->
+      Console.print
+        [ Pp.text
+            "No browser could be found, you will have to open the \
+             documentation yourself.\n"
```
Comment (Alizter) [inline bin/ocaml/doc.ml:65]:
```suggestion
             documentation yourself."
```

File: bin/ocaml/doc.ml [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+      , (* First element of argv is the name of the command. *)
+        (cmd_name :: args) @ [ url ] )
+    in
+    match cmd with
+    | Some (cmd, args) ->
+      Proc.restore_cwd_and_execve
+        (Path.to_absolute_filename cmd)
+        args ~env:Env.initial
+    | None ->
+      Console.print
```
Comment (Alizter) [inline bin/ocaml/doc.ml:62]: If this is meant to be a warning, then you can use User_warning.emit instead.

File: bin/ocaml/doc.ml [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    let url = "file://" ^ absolute_toplevel_index_path in
+    let cmd =
+      let open Option.O in
+      let path = Env_path.path Env.initial in
+      let* cmd_name, args =
+        match Platform.OS.value with
+        | Platform.OS.Darwin -> Some ("open", [ "-u" ])
+        | Platform.OS.Linux -> Some ("xdg-open", [])
+        | Platform.OS.Windows -> Some ("start", [])
+        | Platform.OS.Other -> None
```
Comment (Alizter) [inline bin/ocaml/doc.ml:49]:
```suggestion
        match (Platform.OS.value : Platform.OS.t) with
        | Darwin -> Some ("open", [ "-u" ])
        | Linux -> Some ("xdg-open", [])
        | Windows -> Some ("start", [])
        | Other -> None
```

**Review by Alizter — COMMENTED** [review]
Just a few more comments.

**Comment by EmileTrotignon** [comment]
Hey @Alizter 
I have pushed a commit that takes your comment into account

File: test/blackbox-tests/test-cases/doc-browser.t/dune-project (hunk 16 above)
Comment (Alizter) [reply test/blackbox-tests/test-cases/doc-browser.t/dune-project:1]: This is now 3.10

File: bin/ocaml/doc.ml [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+
+let term =
+  let+ common = Common.term in
+  let config = Common.init common in
+  let request (setup : Import.Main.build_system) =
+    let dir = Path.(relative root) (Common.prefix_target common ".") in
+    let open Action_builder.O in
+    let+ () =
+      Alias.in_dir
+        ~name:(Dune_engine.Alias.Name.of_string "doc")
```
Comment (emillon) [inline bin/ocaml/doc.ml:23]: Can you add that as `Dune_engine.Alias.doc`? (and change the of_string "doc" in that file to use it)

File: test/blackbox-tests/test-cases/doc-browser.t/dune-project (hunk 16 above)
Comment (emillon) [reply test/blackbox-tests/test-cases/doc-browser.t/dune-project:1]: It doesn't matter. If the test works with 3.7 and is not about a 3.7-versioned feature it's fine as is.

File: bin/ocaml/doc.ml [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+
+    let url = "file://" ^ absolute_toplevel_index_path in
+    let cmd =
+      let open Option.O in
+      let path = Env_path.path Env.initial in
+      let* cmd_name, args =
+        match (Platform.OS.value : Platform.OS.t) with
+        | Darwin -> Some ("open", [ "-u" ])
+        | Linux -> Some ("xdg-open", [])
+        | Windows -> None
```
Comment (emillon) [inline bin/ocaml/doc.ml:48]: what does `webbrowser` do here?

File: bin/ocaml/doc.ml [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+    let toplevel_index_path = Dune_rules.Odoc.Paths.toplevel_index doc_ctx in
+    let absolute_toplevel_index_path =
+      Path.(toplevel_index_path |> build |> to_absolute_filename)
+    in
+    Dune_console.print
+      [ Pp.textf "Docs built. Index can be found here: %s\n"
+          absolute_toplevel_index_path
+      ];
+
+    let url = "file://" ^ absolute_toplevel_index_path in
```
Comment (emillon) [inline bin/ocaml/doc.ml:40]: does `absolute_toplevel_index_path` need to be urlencoded here?

File: bin/ocaml/doc.ml [hunk 23]
```diff
@@ -0,0 +1,70 @@
+open Stdune
+open Import
+
+let doc = "Build and view the documentation of an OCaml project"
+
+let man =
+  [ `S "DESCRIPTION"
+  ; `P
+      {|$(b,dune ocaml doc) builds and then opens the documention of an OCaml project in the users default browser.|}
```
Comment (emillon) [inline bin/ocaml/doc.ml:9]:
```suggestion
      {|$(b,dune ocaml doc) builds and then opens the documentation of an OCaml project in the user's default browser.|}
```

File: bin/ocaml/doc.ml (hunk 21 above)
Comment (EmileTrotignon) [reply bin/ocaml/doc.ml:48]: Something similar, although it uses something very weird to get reload instead of a new tab each time.
The relevant code is here : https://github.com/dbuenzli/webbrowser/blob/master/src/webbrowser.ml

File: bin/ocaml/doc.ml [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+    in
+    match cmd with
+    | Some (cmd, args) ->
+      Proc.restore_cwd_and_execve
+        (Path.to_absolute_filename cmd)
+        args ~env:Env.initial
+    | None ->
+      User_warning.emit
+        [ Pp.text
+            "No browser could be found, you will have to open the \
```
[the remark this answers is not in the document]
Comment (EmileTrotignon) [reply bin/ocaml/doc.ml:64]: I It is printed in both cases

File: bin/ocaml/doc.ml (hunk 22 above)
Comment (EmileTrotignon) [reply bin/ocaml/doc.ml:40]: I have tested a path with a space on my machine (linux with firefox), and it works fine as is.

File: bin/ocaml/doc.ml (hunk 20 above)
Comment (EmileTrotignon) [reply bin/ocaml/doc.ml:23]: There is already an Alias.doc variable, but it has the wrong type.

Comment (emillon) [reply bin/ocaml/doc.ml:23]: This has been moved since then, you can find it in the `Alias0` module.

Comment (emillon) [reply bin/ocaml/doc.ml]: You don't need to change the `Alias` module. This identifier already exists, it is in `Alias0` module https://github.com/ocaml/dune/blob/main/src/dune_rules/alias0.ml

File: bin/ocaml/doc.ml [hunk 25]
```diff
@@ -44,7 +44,7 @@ let term =
         | Darwin -> Some ("open", [ "-u" ])
         | Linux -> Some ("xdg-open", [])
         | Windows -> None
-        | Other -> None
+        | Other | FreeBSD | NetBSD | OpenBSD -> None
```
Comment (Alizter) [inline bin/ocaml/doc.ml:47]: FTR BSD platforms can have open or xdg-open but it's probably not worth adding support for it in this PR.
